lib/p2p/Peer.ts

```diff
@@ -265,16 +270,45 @@ class Peer extends EventEmitter {
         resolve();
       };
 
-      const onTimeout = () => {
+      const onError = (err: Error) => {
         cleanup();
-        reject(new Error('Connection timed out.'));
-      };
 
-      this.socket!.once('connect', onConnect);
+        if (!retry) {
+          this.close();
+          reject(err);
+          return;
+        }
+
+        retryDelay = !retryDelay
+          ? Peer.CONNECTION_RETRIES_MIN_DELAY
+          : Math.min(Peer.CONNECTION_RETIES_MAX_DELAY, retryDelay * 2);
+
+        if (Date.now() - startTime + retryDelay > Peer.CONNECTION_RETIES_MAX_PERIOD) {
+          this.close();
+          reject(new Error(`${retries} connection retries failed`));
+          this.emit('ban');
+          return;
+        }
+
+        this.logger.info(
+          `Connection attempt #${retries + 1} to peer (${addressUtils.toString(this.socketAddress)}) ` +
+          `failed: ${err.message}. retrying in ${retryDelay / 1000} sec...`,
+        );
 
-      this.socket!.once('error', onError);
+        setTimeout(() => {
+          retries = retries + 1;
+          this.socket!.connect({ port: this.socketAddress.port, host: this.socketAddress.host });
+          listen();
+        }, retryDelay);
+      };
+
+      const listen = () => {
```

Comment:
"listen" seems a bit inaccurate here, it makes it sound like we're listening for incoming connections. Maybe "bind"?